### PR TITLE
fix: Cleaning up orphaned directories and files when containers creat…

### PR DESCRIPTION
**Pull request by haytok**
…ion fails

When trying to create new containers, the following directories are created based on the new container ID:

  - `/var/lib/nerdctl/1935db59/containers/<namespace>/<container id>`
  - `/var/lib/nerdctl/1935db59/etchosts/<namespace>/<container id>`

If containers with existing names are attempted to be created, the processes fail.
However, in the events of failures, these directories are not cleaned up.

This issue is reported in the following:

 - https://github.com/containerd/nerdctl/issues/2993

This commit resolves the issue by cleaning up the mentioned directories when containers creation fails.

### CODE REVIEW

**Review by djdongjin — COMMENTED** [review]
Thanks.

Is this testable? E.g., can we add a test in https://github.com/containerd/nerdctl/blob/772feb40ee9d0d51db6da8ac3a8f737bbb48475f/cmd/nerdctl/container/container_create_linux_test.go?

File: pkg/dnsutil/hostsstore/hostsstore.go
Comment (apostasie) [reply pkg/dnsutil/hostsstore/hostsstore.go:204]: Agreed they are not necessary.

Intention was to have all methods just follow the same pattern, so that people do not have to think about joining on all returns, whenever things are more complex: https://github.com/containerd/nerdctl/blob/81fe5a47f85eee0c166cd2a7c903a136ab530add/pkg/dnsutil/hostsstore/hostsstore.go#L156-L187

File: pkg/dnsutil/hostsstore/hostsstore.go
Comment (apostasie) [inline pkg/dnsutil/hostsstore/hostsstore.go:199]: @haytok 

Suggesting that:
- we call this method `Delete`, in line with the `safestore.Store` implementation.
- we remove the `DeallocHostsFile` method, and replace the one call to it (which is inside `container.Remove`), by a call to this new `Delete` method

The point being: since we are cleaning up these directories on `failed create`, we should do the same on `remove` now.

LMK what you think.

File: pkg/dnsutil/hostsstore/hostsstore.go
Comment (haytok) [reply pkg/dnsutil/hostsstore/hostsstore.go:204]: Thanks!!! 
With such a simple function, I didn't think it was necessary to process err in defer. So I'll rewrite it to the code suggested by djdongjin.

File: pkg/cmd/container/create.go
Comment (haytok) [reply pkg/cmd/container/create.go:829]: Since it deletes one or more of the following directories, the function name is Dirs.

- `/var/lib/nerdctl/1935db59/containers/<namespace>/<container id>`
- `/var/lib/nerdctl/1935db59/etchosts/<namespace>/<container id>`

File: pkg/dnsutil/hostsstore/hostsstore.go
Comment (haytok) [reply pkg/dnsutil/hostsstore/hostsstore.go:199]: I was thinking the same thing.
The logic to remove `/var/lib/nerdctl/1935db59/etchosts/<namespace>/<container id>/hosts` is called at the following location within `container.Remove()`.

- https://github.com/containerd/nerdctl/blob/main/pkg/cmd/container/remove.go#L222-L228

The cleanup process of `/var/lib/nerdctl/1935db59/etchosts/<namespace>/<container id>`, which is called when container creation fails, is also needed to be  called within `container.Remove()`.

So, when this PR is merged, is it OK to rewrite `hs.DeallocHostsFile(id)` in `container.Remove()` to `hs.DeallocHostsDir(id)` in another PR?

**Comment by haytok** [comment]
Hi, @djdongjin 

Thanks for comments !!!

> Is this testable? E.g., can we add a test in https://github.com/containerd/nerdctl/blob/772feb40ee9d0d51db6da8ac3a8f737bbb48475f/cmd/nerdctl/container/container_create_linux_test.go?

I think it is difficult to implement tests based on the number of files in the directory under `/var/lib/nerdctl/1935db59/` in unit tests.

Is there any E2E test that can perform tests based on the number of local files ... ????

File: pkg/dnsutil/hostsstore/hostsstore.go
Comment (apostasie) [reply pkg/dnsutil/hostsstore/hostsstore.go:199]: They are definitely related changes, so, not shocking to have them together here in this PR.
Or in another PR.
Whatever you are most comfortable with.
:-)

Comment (haytok) [reply pkg/dnsutil/hostsstore/hostsstore.go:199]: > we call this method Delete, in line with the safestore.Store implementation.
we remove the DeallocHostsFile method, and replace the one call to it (which is inside container.Remove), by a call to this new Delete method

I agree with [apostasie](https://github.com/apostasie) that `pkg/dnsutil/hostsstore/hostsstore.go` should be rewritten as follows to implement calling `hs.Delete(id)` within `container.Create()` and `container.Remove()`.

<details><summary>Details</summary>
<p>

```bash
 /Users/haytok/workspace/nerdctl [fix-to-clean-up-orphaned-files]
> gd pkg/dnsutil/hostsstore/hostsstore.go
diff --git a/pkg/dnsutil/hostsstore/hostsstore.go b/pkg/dnsutil/hostsstore/hostsstore.go
index cf691542..507f93ec 100644
--- a/pkg/dnsutil/hostsstore/hostsstore.go
+++ b/pkg/dnsutil/hostsstore/hostsstore.go
@@ -89,8 +89,7 @@ type Store interface {
        Release(id string) error
        Update(id, newName string) error
        HostsPath(id string) (location string, err error)
-       DeallocHostsFile(id string) (err error)
-       DeallocHostsDir(id string) (err error)
+       Delete(id string) (err error)
        AllocHostsFile(id string, content []byte) (location string, err error)
 }

@@ -186,24 +185,13 @@ func (x *hostsStore) AllocHostsFile(id string, content []byte) (location string,
        return x.safeStore.Location(id, hostsFile)
 }

-func (x *hostsStore) DeallocHostsFile(id string) (err error) {
-       defer func() {
-               if err != nil {
-                       err = errors.Join(ErrHostsStore, err)
-               }
-       }()
-
-       return x.safeStore.WithLock(func() error { return x.safeStore.Delete(id, hostsFile) })
-}
-
-func (x *hostsStore) DeallocHostsDir(id string) (err error) {
-       defer func() {
-               if err != nil {
-                       err = errors.Join(ErrHostsStore, err)
-               }
-       }()
+func (x *hostsStore) Delete(id string) (err error) {
+       err = x.safeStore.WithLock(func() error { return x.safeStore.Delete(id) })
+       if err != nil {
+               err = errors.Join(ErrHostsStore, err)
+       }

-       return x.safeStore.WithLock(func() error { return x.safeStore.Delete(id) })
+       return err
 }

 func (x *hostsStore) HostsPath(id string) (location string, err error) {
```

</p>
</details>

Comment (haytok) [reply pkg/dnsutil/hostsstore/hostsstore.go]: Fixed

File: pkg/dnsutil/hostsstore/hostsstore.go
Comment (haytok) [reply pkg/dnsutil/hostsstore/hostsstore.go]: Fixed

**Comment by apostasie** [comment]
> Is there any E2E test that can perform tests based on the number of local files ... ????

Here is an e2e test for you that does it (the tricky part is to access the right location):

```go
func TestBug(t *testing.T) {
	nerdtest.Setup()

	testCase := &test.Case{
		Description: "Issue #2993 - nerdctl is no longer leaking etchosts dirs & files on failure",
		// Make sure we are in a private env so that other tests do not pollute this
		Require: nerdtest.Private,
		// Run a first container successfully
		Setup: func(data test.Data, helpers test.Helpers) {
			helpers.Ensure("run", "--name", data.Identifier(), testutil.CommonImage)
		},
		Cleanup: func(data test.Data, helpers test.Helpers) {
			helpers.Anyhow("rm", "-f", data.Identifier())
		},
		// Try to run a second container with the same name
		Command: func(data test.Data, helpers test.Helpers) test.Command {
			return helpers.Command("run", "--name", data.Identifier(), testutil.CommonImage)
		},
		Expected: func(data test.Data, helpers test.Helpers) *test.Expected {
			return &test.Expected{
				// We expect the command to fail
				ExitCode: 1,
				// As the name is already used
				Errors: []error{errors.New("is already used by")},
				Output: func(stdout string, info string, t *testing.T) {
					// Get the test DataRoot, the containerd address, and get the dataStore location
					base := testutil.NewBase(t)
					dataStore, err := clientutil.DataStore(string(data.ReadConfig(nerdtest.DataRoot)), base.ContainerdAddress())
					assert.NilError(t, err, "data store failed")
					// Now, we are interested in "etchosts", and in the Namespace (data.Identifier() with `nerdtest.Private`)
					etchostsdir := filepath.Join(dataStore, "etchosts", data.Identifier())
					// Read the dir
					files, err := os.ReadDir(etchostsdir)
					assert.NilError(t, err, fmt.Sprintf("failed reading %s", etchostsdir))
					// We are expecting only one directory there, as the other container should have deleted its directory on failure
					assert.Equal(t, len(files), 1, "there should be only 1 subdirectory in the etchosts store")
				},
			}
		},
	}

	testCase.Run(t)
}

```

**Comment by haytok** [comment]
Thanks !!!!!!!
I'll add test for this bug.

**Comment by apostasie** [comment]
Pending addition of a test validating the fix, LGTM.

CI failures are unrelated (windows regression #3437).

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (apostasie) [inline cmd/nerdctl/container/container_create_linux_test.go:185]: You need to mark this test as DockerIncompatible.

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (apostasie) [inline cmd/nerdctl/container/container_create_linux_test.go:186]: Suggesting you use a dedicated namespace for that test, so that we can parallelize the test.

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (haytok) [reply cmd/nerdctl/container/container_create_linux_test.go:185]: Thanks !
Added `testutil.DockerIncompatible(t)` to tests.

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (haytok) [reply cmd/nerdctl/container/container_create_linux_test.go]: Thanks, fixed to specify independent namespaces for each test case.

**Comment by haytok** [comment]
Hi, @apostasie 
Thanks for the review.
I have added tests for this issue, so could you re-review when you have time ?
CI failure is related to this comment (https://github.com/containerd/nerdctl/pull/3457#issuecomment-2373055605)

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (apostasie) [inline cmd/nerdctl/container/container_create_linux_test.go:209]: Use `testutil.Identifier(t)` for the namespace name.
Same thing for container name.

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (apostasie) [inline cmd/nerdctl/container/container_create_linux_test.go:211]: Use `t.Cleanup` instead of defer.
Also, be sure to also call your cleanup routine before starting.

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (apostasie) [inline cmd/nerdctl/container/container_create_linux_test.go:248]: You also want to make sure you run this ^ before anything else, so that leftovers from previous tests do not make you fail.

**Review by apostasie — CHANGES_REQUESTED** [review]
Thanks @haytok !

Left a few comments on tests.

If you want, there is more info about tests principles here: https://github.com/containerd/nerdctl/blob/main/docs/testing/README.md#principles

And details about a new tooling framework here:
https://github.com/containerd/nerdctl/blob/main/docs/testing/tools.md

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (apostasie) [inline cmd/nerdctl/container/container_create_linux_test.go:211]: Use t.Cleanup instead of defer.
Also make sure to call your cleanup routine ALSO before you start doing anything to take care of leftovers from previous tests.

**Comment by haytok** [comment]
The local branch was out of date and I was unaware that there was documentation for the test. 
Thanks for sharing!
I will check the documentation and fix it.

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (haytok) [reply cmd/nerdctl/container/container_create_linux_test.go]: Thanks !!! fixed based on [the testing principles](https://github.com/containerd/nerdctl/blob/main/docs/testing/tools.md).

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (haytok) [reply cmd/nerdctl/container/container_create_linux_test.go]: Thanks !!! Fixed.

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (haytok) [reply cmd/nerdctl/container/container_create_linux_test.go]: fixed based on [the testing principles](https://github.com/containerd/nerdctl/blob/main/docs/testing/tools.md).

**Review by djdongjin — APPROVED** [review]
LGTM, thanks

File: cmd/nerdctl/container/container_create_linux_test.go
Comment (haytok) [reply cmd/nerdctl/container/container_create_linux_test.go:194]: Thanks! I'll fix!

Comment (haytok) [reply cmd/nerdctl/container/container_create_linux_test.go]: fixed

**Comment by apostasie** [comment]
@haytok can you [squash](https://github.com/farcloser/open-source-help-desk/blob/main/GIT.md#how-to-squash)?

Reviewers, about the CI failures:
- https://github.com/containerd/nerdctl/actions/runs/11074391386/job/30773073228?pr=3457#step:6:537 - `container_run_mount_linux_test.go:367: assertion failed: error is not nil: invalid /proc/mounts format ""` - no idea :/ - fluke?
- https://github.com/containerd/nerdctl/actions/runs/11074391386/job/30773072030?pr=3457#step:6:2968
   - this is #3396
- https://github.com/containerd/nerdctl/actions/runs/11074391387/job/30773073216?pr=3457#step:9:568
   - this is #3437

**Comment by haytok** [comment]
Hi, @apostasie 

I have squashed all commits.

**Review by apostasie — APPROVED** [review]

**Comment by apostasie** [comment]
@haytok appreciate all your efforts here!
Looks good.
Now to maintainers.
@djdongjin @AkihiroSuda

**Pull request merged**
